File: services/server/src/safe_socket/safe_socket_test.py
import struct
import unittest

from safe_socket import send_frame, recv_frame


class _FakeSocket:

    _MAX_CHUNK = 3

    def __init__(self, initial_bytes: bytes = b""):
        self._inbox = bytearray(initial_bytes)
        self.sent = bytearray()

    def send(self, data: bytes) -> int:
        chunk = data[: self._MAX_CHUNK]
        self.sent.extend(chunk)
        return len(chunk)

    def recv(self, size: int) -> bytes:
        to_read = min(self._MAX_CHUNK, size)
        chunk = bytes(self._inbox[:to_read])
        del self._inbox[:to_read]
        return chunk


class SendFrameTest(unittest.TestCase):
    def test_send_frame_writes_length_prefixed_payload_despite_short_writes(self):
        sock = _FakeSocket()

        send_frame(sock, b"hello")

        self.assertEqual(bytes(sock.sent), struct.pack(">I", 5) + b"hello")


class RecvFrameTest(unittest.TestCase):
    def test_recv_frame_reads_length_prefixed_payload_despite_short_reads(self):
        sock = _FakeSocket(struct.pack(">I", 8) + b"abcdefgh")

        payload = recv_frame(sock)

        self.assertEqual(payload, b"abcdefgh")

    def test_recv_frame_returns_empty_payload_for_zero_length_frame(self):
        sock = _FakeSocket(struct.pack(">I", 0))

        payload = recv_frame(sock)

        self.assertEqual(payload, b"")


if __name__ == "__main__":
    unittest.main()

File: services/client/src/safe_socket/safe_socket_test.go
package safe_socket

import (
	"bytes"
	"io"
	"testing"
)

type shortReader struct {
	data     []byte
	maxChunk int
}

func (r *shortReader) Read(p []byte) (int, error) {
	if len(r.data) == 0 {
		return 0, io.EOF
	}
	n := r.maxChunk
	if n > len(r.data) {
		n = len(r.data)
	}
	if n > len(p) {
		n = len(p)
	}
	copy(p, r.data[:n])
	r.data = r.data[n:]
	return n, nil
}

type shortWriter struct {
	written  []byte
	maxChunk int
}

func (w *shortWriter) Write(p []byte) (int, error) {
	n := w.maxChunk
	if n > len(p) {
		n = len(p)
	}
	w.written = append(w.written, p[:n]...)
	return n, nil
}

func TestSendFrame_WritesLengthPrefixedPayloadDespiteShortWrites(t *testing.T) {
	writer := &shortWriter{maxChunk: 3}

	if err := SendFrame(writer, []byte("hello")); err != nil {
		t.Fatalf("SendFrame returned error: %v", err)
	}

	want := []byte{0, 0, 0, 5, 'h', 'e', 'l', 'l', 'o'}
	if !bytes.Equal(writer.written, want) {
		t.Fatalf("SendFrame wrote %v, want %v", writer.written, want)
	}
}

func TestRecvFrame_ReadsLengthPrefixedPayloadDespiteShortReads(t *testing.T) {
	reader := &shortReader{
		data:     []byte{0, 0, 0, 8, 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'},
		maxChunk: 3,
	}

	payload, err := RecvFrame(reader)
	if err != nil {
		t.Fatalf("RecvFrame returned error: %v", err)
	}

	want := []byte("abcdefgh")
	if !bytes.Equal(payload, want) {
		t.Fatalf("RecvFrame() = %v, want %v", payload, want)
	}
}

func TestRecvFrame_ReturnsEmptyPayloadForZeroLengthFrame(t *testing.T) {
	reader := &shortReader{data: []byte{0, 0, 0, 0}, maxChunk: 3}

	payload, err := RecvFrame(reader)
	if err != nil {
		t.Fatalf("RecvFrame returned error: %v", err)
	}

	if len(payload) != 0 {
		t.Fatalf("RecvFrame() = %v, want empty", payload)
	}
}
